### PR TITLE
storage: use critical handles in the storage controller 

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

This PR integrates persist critical handles in the storage controller to avoid having to rely on read leases for correctness. The new critical handle abstraction requires that whenever a critical handle is greated its ID must be durably persisted and remembered. If it gets lost the associated shard will be "wedged" forever.

The storage controller maintains a single critical handle per collection and only for the `data_shard`. To avoid durably persisting handle ids it uses the same hardcoded critical handle reader id (`mz_persist_client::PersistClient::CONTROLLER_CRITICAL_SINCE`) to manage upper bound of the since frontier for the collections it manages.

With critical handles it is crucial that correct fencing is implemented so that there is only one leader in charge and all other processes are fenced off from interfering. Since handles support an opaque token that has compare and exchange semantics. In the storage controller we store the `environmentd` epoch to fence off instances of the storage controller with a lower epoch. This epoch is durably persisted by `environmentd` and it is being atomically incremented on every startup. When a fence error is received then we halt the process as that indicates that another envd is running somewhere.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/storage-client/src/controller.rs
Comment (danhhz) [inline src/storage-client/src/controller.rs:1226]: genuine question (I don't know): is there any value in sending this to the worker vs just calling register/snap/expire here?

File: src/storage-client/src/controller.rs
Comment (danhhz) [inline src/storage-client/src/controller.rs:1841]: nit: I'd mention _why_ (so we don't have to keep heartbeating a handle just to occasionally use it for a snapshot)

File: src/storage-client/src/controller.rs
Comment (danhhz) [inline src/storage-client/src/controller.rs:1882]: IIUC the token here is meant to be the envd epoch, but you'll want to talk to @jkosh44 about where that lives and how we should plumb it down (probably can be a constructor arg to `persist_read_handles::PersistWorker`)

**Review by danhhz — APPROVED** [review]
LGTM modulo getting Joe's signoff on the compare_and_downgrade token

File: src/storage-client/src/controller.rs
Comment (petrosagg) [reply src/storage-client/src/controller.rs:1226]: Good question! I don't know why it was written that way but I agree that there is no point shipping the command to another async worker only to await its results. I'll move the logic in here since I'm working in this area

File: src/storage-client/src/controller.rs
Comment (jkosh44) [reply src/storage-client/src/controller.rs:1882]: Currently environmentd creates the `mz_controller::Controller`, with the epoch here:
https://github.com/MaterializeInc/materialize/blob/aaf3558f4e983cf5b432bcd2af35982cdc0c2c89/src/environmentd/src/lib.rs#L278-L279

The `mz_controller::Controller` creates the `mz_storage_client::controller::Controller` here: https://github.com/MaterializeInc/materialize/blob/aaf3558f4e983cf5b432bcd2af35982cdc0c2c89/src/controller/src/lib.rs#L213-L224

The `mz_storage_client::controller::Controller` creates the `StorageControllerState` here:
https://github.com/MaterializeInc/materialize/blob/aaf3558f4e983cf5b432bcd2af35982cdc0c2c89/src/storage-client/src/controller.rs#L1381-L1408

and finally the `StorageControllerState` creates the `persist_read_handles::PersistWorker` here:
https://github.com/MaterializeInc/materialize/blob/aaf3558f4e983cf5b432bcd2af35982cdc0c2c89/src/storage-client/src/controller.rs#L738-L771

So you should be able to easily plumb the epoch down through all those constructors if I followed everything correctly.

Comment (jkosh44) [reply src/storage-client/src/controller.rs:1882]: If you do need to access the epoch in other places, it's actually a bit hidden. It's made available through the stash API here: 
https://github.com/MaterializeInc/materialize/blob/aaf3558f4e983cf5b432bcd2af35982cdc0c2c89/src/stash/src/lib.rs#L396-L398

You'll likely need to add a method to the `Connection` object that exposes this:
https://github.com/MaterializeInc/materialize/blob/aaf3558f4e983cf5b432bcd2af35982cdc0c2c89/src/adapter/src/catalog/storage.rs#L529
Something like 
```Rust
    pub fn epoch(&self) -> Result<Option<i64>, Error> {
        Ok(self.stash.epoch().await?)
    }
```

Then you'll be able use that method in the Coordinator to pass it to the storage controller/client like so:
```Rust
pub async fn some_coord_method(&self) {
  let epoch = self.catalog.storage().await.epoch()?;
  // do something with epoch.
}
```

Comment (petrosagg) [reply src/storage-client/src/controller.rs:1882]: This was super helpful, thanks! I will wire it up tomorrow

**Comment by jkosh44** [comment]
> LGTM modulo getting Joe's signoff on the compare_and_downgrade token

Just to clarify, did you want me to signoff on how the token is being used in the storage controller or just how to plumb the token down into the storage controller?

File: src/storage-client/src/controller.rs
Comment (danhhz) [inline src/storage-client/src/controller.rs:918]: we want to switch this to the non "maybe_" version, right?

**Review by danhhz — APPROVED** [review]
> Just to clarify, did you want me to signoff on how the token is being used in the storage controller or just how to plumb the token down into the storage controller?

ah sorry, my message was ambiguous. it was more the latter, though you're certainly welcome to poke at the usage if you like. the protocol is just that larger epochs fence out smaller epochs, correct? also, the persist code this is using currently assumes that we never see an epoch smaller than i64::default(), is that an okay assumption?

File: src/storage-client/src/controller.rs
Comment (petrosagg) [reply src/storage-client/src/controller.rs:918]: That's right

File: src/storage-client/src/controller.rs
Comment (danhhz) [inline src/storage-client/src/controller.rs:926]: ah, looks like there might be some race condition here, where the epoch is not whatever we got at registration time, but also is still < our_epoch. I guess just toss this whole bit in a loop?

**Comment by jkosh44** [comment]
> the protocol is just that larger epochs fence out smaller epochs, correct?

Yes that's correct.

> the persist code this is using currently assumes that we never see an epoch smaller than i64::default(), is that an okay assumption?

I don't think this is a valid assumption. Currently in the code I believe that the epoch always happens to be greater than or equal to 0, but the API make no guarantee that it will never be negative. @mjibson would you agree?

File: src/storage-client/src/controller.rs
Comment (petrosagg) [reply src/storage-client/src/controller.rs:926]: The halt in the CI happens further down which is why it's so weird. It just has an identical message but you can see the module path The halt happening there means that envd successfully went past this section, which takes ownership of a since handle, but then someone flipped the epoch back to zero!

File: src/persist-client/src/critical.rs
Comment (petrosagg) [inline src/persist-client/src/critical.rs:379]: @danhhz @pH14 I checked in this test which I came up while trying to debug the CI failures and this assert fires. Is this expected?

Comment (pH14) [reply src/persist-client/src/critical.rs:379]: hm, right now we only update the `since` and `opaque` if `PartialOrder::less_than(state_since, new_since)` but I wonder if that ought to be `less_equal` instead

Comment (pH14) [reply src/persist-client/src/critical.rs:379]: yeah I think that should be fine. I've got a few small tweaks incl this one I can PR

File: src/storage-client/src/controller.rs
Comment (mjibson) [reply src/storage-client/src/controller.rs:926]: That sounds like something is dropping the stash tables and recreating them. Another test running in parallel?

**Comment by mjibson** [comment]
> I don't think this is a valid assumption. Currently in the code I believe that the epoch always happens to be greater than or equal to 0, but the API make no guarantee that it will never be negative. @mjibson would you agree?

Correct. It happens to start at 0 today, but the API does not make that guarantee. Using `i64::default` not recommended.

File: src/persist-client/src/critical.rs
Comment (pH14) [reply src/persist-client/src/critical.rs:379]: https://github.com/MaterializeInc/materialize/pull/15989

**Comment by mjibson** [comment]
Looking at [the code](https://github.com/MaterializeInc/materialize/blob/ad02fffb6f399be870a304a391e1089615ced5d9/src/persist-client/src/internal/state.rs#L224) for this: it's definitely not safe to mint one's own epoch and start comparing against that. The `epoch` method in stash is only safe if it is only compared to other calls against `epoch`. Using `i64::MIN` is also unsafe as a "lowest starting number to compare against". This is because the stash also uses a nonce when determining its leadership to prevent against the case of two separate stashes where one drops the stash and starts over also at 0.

**Comment by danhhz** [comment]
okay this is exactly the sort of thing I was hoping to shake out this week! we can certainly do something where either (1) on the first registration, we set epoch to None and it matches any expected epoch or (2) we take an epoch as part of registration, but both of those are pretty clunky and I dislike them

the simplest thing would be to change the coord epoch protocol such that we can do this default thing, is that on the table?  we could also swap `Default` for a new trait that has to be explicitly implemented (think `Timestamp` and `minimum`) so we could use something like `-1` if that would work

**Comment by danhhz** [comment]
> we could also swap Default for a new trait that has to be explicitly implemented (think Timestamp and minimum) so we could use something like -1 if that would work
oh whoops. you already answered that, sorry!

**Comment by danhhz** [comment]
another idea is something that was discussed in the storage hangout yesterday, which is not using the coord epoch here and instead inventing our own fencing token inside the storage controller stash. but I have nowhere near the context I'd need to have an opinion on whether that's a good idea or not

**Comment by danhhz** [comment]
> I hope this get to the right place as I'm replying from email. If we could
have the token be None on the first registration would be amazing!

oh ha! concurrent with this comment, matt and I just had a huddle where we decided we liked the trait option, and I went and typed that up: https://github.com/MaterializeInc/materialize/pull/15997. matt was also bullish on PR (followup PR to this one) that switches from returning the coord epoch as an i64 to a `Epoch` struct (like we do with timestamps) so that the logic (e.g which epoch "wins", currently being duplicated into this PR) can be encapsulated in one place. my proposed trait PR would also let us do the same for the logic the initial value logic

**Review by danhhz — APPROVED** [review]

File: src/storage-client/src/controller.rs
Comment (guswynn) [inline src/storage-client/src/controller.rs:914]: this +1 is a bit weird, i know we can't use the epoch itself because we need to succeed a `compare_and_downgrade_since` against the default epoch of the first since handle, but explaining the +1 here might be nice

File: src/storage-client/src/controller.rs
Comment (guswynn) [inline src/storage-client/src/controller.rs:1831]: very nice! Though, this could, in extremely rare circumstances, starve the code below right?

**Review by guswynn — APPROVED** [review]
:seemsreasonable:

just had a few non-urgent questions

File: src/storage-client/src/controller.rs
Comment (mjibson) [inline src/storage-client/src/controller.rs:915]: I agree above with the `+1` being weird. I'm not convinced by the use of `<` here either.

Comment (petrosagg) [reply src/storage-client/src/controller.rs:915]: Well, I was about to explain my reasoning and then it didn't make sense! I'll leave the epoch as-is and make this a `<=`

Comment (petrosagg) [reply src/storage-client/src/controller.rs:915]: I guess there one argument for the `<` which is that if we ever fuck up and have the same epoch in two machines one will crash instead of both happily continuing.

File: src/storage-client/src/controller.rs
Comment (petrosagg) [reply src/storage-client/src/controller.rs:1831]: It's hard to imagine this actually happening but yeah. It's the same logic as before though, I just moved the code around

Comment (guswynn) [reply src/storage-client/src/controller.rs:1831]: yeah we can leave this for now! At some point we may want to add metrics on this channel!

File: src/storage-client/src/controller.rs
Comment (guswynn) [reply src/storage-client/src/controller.rs:915]: > have the same epoch in two machines one will crash instead of both happily continuing.

This is not something we should defend against, its a way bigger bug than the controller!

File: src/storage-client/src/controller.rs
Comment (mjibson) [reply src/storage-client/src/controller.rs:914]: > we need to succeed a compare_and_downgrade_since against the default epoch of the first since handle

The default epoch is 0 right? The envd epoch function will current never return anything < 1, so I think it is ok to not add one? (This is currently not a documented guarantee, but it can be.)

File: src/stash/src/lib.rs
Comment (mjibson) [inline src/stash/src/lib.rs:398]:
```suggestion
    /// Returns the stash's epoch number. If `Some`, it is a positive number that
    /// increases with each start of a stash.
``

We can increase the documented guarantees further here too.

**Review by mjibson — APPROVED** [review]
First commit LGTM, although I don't understand how non zero helps here.

Can you point me to the code that implements `Codec64` for `Option<NonZeroI64>`? I'm curious how that works, and I don't anything interesting listed under Implementors at https://dev.materialize.com/api/rust-private/mz_persist_types/trait.Codec64.html#implementors.

The commit message says `Option<Epoch>`. Does `Epoch` there mean `NonZeroI64`, since that's the type that lives in `ComputeStartupEpoch`?

**Comment by petrosagg** [comment]
@mjibson Here you are https://github.com/MaterializeInc/materialize/blob/93caa48691612906644b2b5316e6ed7bca06b64d/src/storage-client/src/controller.rs#L1452-L1476

**Comment by danhhz** [comment]
\o/ Thanks for seeing this through!

**Pull request merged**
